### DISCUSSION TITLE
oauth response object

### DISCUSSION THREAD

**Discussion by martino87r** [fastapi-users/fastapi-users#1104] [Q&A]
Hi there, I've implemented the oauth (google) backend and would like to know how exactly should it be implemented in the client...

Let me clarify:

- The client initiates the authorization request to obtain the authentication URL
- The client opens a popup (`window.open`...) and invokes the authorization URL
- The user either accepts or denies access to my application to google user data
- Google redirects to the callback URL
- My the response object is JSON thus cannot (at least for me) notify the parent window it received a valid (or invalid) auth_token

In my mind the simplest fix would be that the response would be an HTML page with a bit of JS to post and notify the token back to the main window, ideally using `window.dispatchEvent`. Any ideas or suggestions?

**Comment by frankie567** [comment]
What I usually do in those cases is to make the redirect URL pointing to an URL on my frontend. This way, I can have JS code catch the authorization code in the query parameters and make an XHR request to my /callback API. Then, the frontend directly gets the response of the callback and can use/store the token.

**Reply by martino87r** [reply]
I think that would complicate the flow a whole lot, can you even trust the authorization data coming from a frontend once you start processing the response and issuing for example a server side session?

**Reply by frankie567** [reply]
Yes, you can, that's the point of OAuth2. All you need to do is to get the authorization code from the query parameters (`code`) and send it to the `/callback` route. Then, the server will exchange with Google this authorization code to get proper credentials from them. Your frontend is just passing values here.

**Reply by martino87r** [reply]
Ahh now it makes sense indeed :) Going to give it a shot thanks for the help
